### ISSUE TITLE
[Help] How to free up memory after each segmentation

### ISSUE THREAD

**Issue by xshang93** [facebookresearch/segment-anything#384]
Hi, I'm using SAM to get masks for segmenting my images (~320, 960*800 ea.) on CPU, but I noticed that SAM eats up the memory quickly. I'm using a for loop to segment each image, and after ~60 iterations the memory usage adds up gradually to 6GBs, which max out the free memory in my desktop. I wonder if there is a way to release memory after each iteration is finished?

My code looks like this (My for loop calls this function in each iteration):
```
def mp_ext(img):
    .... variables are defined here and removed since it's irrelevant to my question
    # Tell SAM to remember this image
    image = cv2.imread(img)
    predictor.set_image(image)
    # Points are input to the model in (x,y) format and come with labels 1 (foreground point) or 0 (background point)
    input_point = np.array([centre])
    input_label = np.array([1])
    
    # Make prediction
    masks, scores, logits = predictor.predict(
        point_coords=input_point,
        point_labels=input_label,
        multimask_output=True,
    )
    mask_input = logits[np.argmax(scores), :, :]  # Choose the model's best mask
    
    # add extra points for the melt track
    input_point = np.array([p1,p2,p3,p4])
    input_label = np.array([1,1,1,1])
    
    masks, _, _ = predictor.predict(
        point_coords=input_point,
        point_labels=input_label,
        mask_input=mask_input[None, :, :],
        multimask_output=False,
    )
```

Thank you!

**Comment by urimerhav** [comment]
+1 on this. also experienced OOM after running many iterations.

**Comment by fsemerar** [comment]
I think this is what predictor.reset_image() is for

**Comment by xshang93** [comment]
> I think this is what predictor.reset_image() is for

Unfortunately I wasn't able to get this to work by adding `redictor.reset_image()` after each run. Maybe I'm doing something wrong? Can you please confirm if you were able to make this work?

**Comment by lavishchauhan321** [comment]
Using torch.cuda.empty_cache() after every function call should work

**Comment by nikste** [comment]
using torch.cuda.empty_cache() didn't work for me unfortunately, deleting all the python objects created in the function neither (aside from reloading the model) it seems to be related to the predictor (also tried the reset function)..

**Comment by AlinGrigoras** [comment]
The solution is to use a multiprocessing approach, so you should move the image processing task to a separate process. By doing so, you isolate the memory usage of the task from the main process. This ensures that any memory consumed during processing is released back to the system when the process terminates, effectively preventing the memory leak

**Comment by JZ-mcovision** [comment]
same problem
